File: php_legacy/src/Services/SqlValidator.php
<?php
declare(strict_types=1);

/**
 * SqlValidator
 * ============
 *
 * Multi-layer safety gate for SQL produced by an LLM.
 * LLM output is NEVER trusted raw — this validator must pass before exec.
 *
 * Guardrails (fail closed):
 *   1. Non-empty after trim
 *   2. Single statement only (no stacked queries via ';')
 *   3. SELECT / WITH (CTE) only — no DDL or DML
 *   4. No forbidden keywords (DROP, ALTER, INSERT, UPDATE, DELETE,
 *      TRUNCATE, RENAME, GRANT, REVOKE, CREATE, LOAD, HANDLER, LOCK)
 *   5. No multi-statement tricks (comments, /*... *\/, -- )
 *   6. Only whitelisted tables referenced
 *   7. LIMIT injected if missing, capped at MAX_ROWS
 *
 * Any failure throws InvalidArgumentException with a reason suitable for
 * logging but NOT for direct display to end users (information leak risk).
 */
final class SqlValidator
{
    public const MAX_ROWS = 1000;

    /** Whitelist of tables the LLM is allowed to reference. */
    private const ALLOWED_TABLES = [
        'fact_allotment',
        'dim_iit',
        'dim_branch',
        'dim_quota',
        'dim_seat_type',
        'dim_gender',
    ];

    /**
     * Forbidden keywords, compiled as word-boundary regex.
     * We treat even read-only `SHOW`/`DESCRIBE` as forbidden — the API is
     * meant for data questions, not metadata introspection.
     */
    private const FORBIDDEN = [
        'INSERT', 'UPDATE', 'DELETE', 'DROP', 'ALTER', 'CREATE', 'TRUNCATE',
        'RENAME', 'GRANT', 'REVOKE', 'REPLACE', 'MERGE', 'CALL', 'EXEC',
        'EXECUTE', 'PREPARE', 'DEALLOCATE', 'HANDLER', 'LOCK', 'UNLOCK',
        'LOAD', 'OUTFILE', 'DUMPFILE', 'INTO\s+OUTFILE', 'INTO\s+DUMPFILE',
        'SET\s+GLOBAL', 'SET\s+SESSION', 'SHOW', 'DESCRIBE', 'DESC\s',
        'EXPLAIN', 'USE\s+', 'BENCHMARK', 'SLEEP',
    ];

    /**
     * Validate and rewrite SQL. Returns the safe SQL to execute.
     *
     * @throws InvalidArgumentException if any guardrail fails
     */
    public static function validateAndPrepare(string $sql): string
    {
        $sql = trim($sql);

        if ($sql === '') {
            throw new InvalidArgumentException('Empty SQL');
        }

        // Strip trailing semicolons (single one is fine, multiple = stacking)
        $sql = rtrim($sql, "; \t\n\r\0\x0B");

        // ---------- Layer 1: no stacked statements ----------
        // After trimming the final semicolon, if there's still a ';' anywhere
        // outside a quoted string, it's a stacked-statement attack.
        if (self::containsUnquotedSemicolon($sql)) {
            throw new InvalidArgumentException('Multiple statements not allowed');
        }

        // ---------- Layer 2: remove comments, then re-check ----------
        $stripped = self::stripComments($sql);

        // ---------- Layer 3: must START with SELECT or WITH ----------
        if (!preg_match('/^\s*(SELECT|WITH)\b/i', $stripped)) {
            throw new InvalidArgumentException('Only SELECT / WITH queries are allowed');
        }

        // ---------- Layer 4: forbidden keywords ----------
        foreach (self::FORBIDDEN as $kw) {
            if (preg_match('/\b' . $kw . '\b/i', $stripped)) {
                throw new InvalidArgumentException("Forbidden keyword detected: {$kw}");
            }
        }

        // ---------- Layer 5: table whitelist ----------
        // Grab every token following FROM or JOIN, lowercase it, check against whitelist.
        $tables = self::extractTables($stripped);
        foreach ($tables as $t) {
            if (!in_array($t, self::ALLOWED_TABLES, true)) {
                throw new InvalidArgumentException("Disallowed table: {$t}");
            }
        }

        // ---------- Layer 6: LIMIT injection ----------
        // If LLM provided a LIMIT, cap it; if not, append one.
        $final = self::enforceLimit($sql, self::MAX_ROWS);

        return $final;
    }

    private static function containsUnquotedSemicolon(string $sql): bool
    {
        $len = strlen($sql);
        $inSingle = false; $inDouble = false; $inBacktick = false;

        for ($i = 0; $i < $len; $i++) {
            $ch = $sql[$i];
            $prev = $i > 0 ? $sql[$i - 1] : '';

            if ($ch === "'" && $prev !== '\\' && !$inDouble && !$inBacktick) {
                $inSingle = !$inSingle;
            } elseif ($ch === '"' && $prev !== '\\' && !$inSingle && !$inBacktick) {
                $inDouble = !$inDouble;
            } elseif ($ch === '`' && !$inSingle && !$inDouble) {
                $inBacktick = !$inBacktick;
            } elseif ($ch === ';' && !$inSingle && !$inDouble && !$inBacktick) {
                return true;
            }
        }
        return false;
    }

    private static function stripComments(string $sql): string
    {
        // Remove /* ... */ comments (non-greedy)
        $sql = preg_replace('!/\*.*?\*/!s', ' ', $sql);
        // Remove -- line comments
        $sql = preg_replace('/--[^\n]*/', ' ', $sql);
        // Remove # line comments (MySQL-specific)
        $sql = preg_replace('/#[^\n]*/', ' ', $sql);
        return $sql;
    }

    /**
     * Extract table names after FROM / JOIN, excluding:
     *   - subquery aliases:  FROM (SELECT ...) t
     *   - CTE aliases:       WITH foo AS (SELECT ...) ... FROM foo
     *
     * Strips backticks and schema prefixes.
     */
    private static function extractTables(string $sql): array
    {
        // 1. Collect CTE names declared in WITH clauses. They're aliases, not real tables.
        $cteNames = [];
        if (preg_match_all('/\bWITH\s+(.+?)\s+SELECT\b/is', $sql, $withMatches)) {
            foreach ($withMatches[1] as $withBody) {
                // Match each "name AS (" occurrence
                if (preg_match_all('/\b([a-zA-Z_][a-zA-Z0-9_]*)\s+AS\s*\(/i', $withBody, $cteMatches)) {
                    foreach ($cteMatches[1] as $name) {
                        $cteNames[] = strtolower($name);
                    }
                }
            }
        }

        // 2. Find tables after FROM/JOIN, but skip if followed by '(' (i.e. subquery)
        $tables = [];
        $pattern = '/\b(?:FROM|JOIN)\s+`?([a-zA-Z_][a-zA-Z0-9_]*)`?/i';

        if (preg_match_all($pattern, $sql, $matches)) {
            foreach ($matches[1] as $t) {
                $lower = strtolower($t);
                // Skip CTE aliases — they're not real tables
                if (in_array($lower, $cteNames, true)) continue;
                $tables[] = $lower;
            }
        }
        return array_unique($tables);
    }

    private static function enforceLimit(string $sql, int $max): string
    {
        // Check for LIMIT with optional OFFSET
        // Matches: LIMIT 10
        // Matches: LIMIT 10, 20
        // Matches: LIMIT 10 OFFSET 20
        if (preg_match('/\bLIMIT\s+(\d+)(?:\s*(?:,|OFFSET)\s*(\d+))?\s*$/i', $sql, $m)) {
            $isOffsetSyntax = stripos($m[0], 'OFFSET') !== false;
            $hasComma = strpos($m[0], ',') !== false;

            if ($isOffsetSyntax) {
                $count = (int) $m[1];
            } elseif ($hasComma) {
                $count = (int) $m[2];
            } else {
                $count = (int) $m[1];
            }

            if ($count > $max) {
                if ($isOffsetSyntax) {
                    return preg_replace(
                        '/\bLIMIT\s+(\d+)\s+OFFSET\s+(\d+)\s*$/i',
                        "LIMIT {$max} OFFSET $2",
                        $sql
                    );
                } elseif ($hasComma) {
                    return preg_replace(
                        '/\bLIMIT\s+(\d+)\s*,\s*(\d+)\s*$/i',
                        "LIMIT $1, {$max}",
                        $sql
                    );
                } else {
                    return preg_replace(
                        '/\bLIMIT\s+(\d+)\s*$/i',
                        "LIMIT {$max}",
                        $sql
                    );
                }
            }
            return $sql;
        }

        // No LIMIT → append one
        return $sql . " LIMIT {$max}";
    }
}

File: analytics/services/sql_validator.py
import re


class SqlValidator:
    MAX_ROWS = 1000

    ALLOWED_TABLES = {
        'fact_allotment',
        'dim_iit',
        'dim_branch',
        'dim_quota',
        'dim_seat_type',
        'dim_gender',
    }

    FORBIDDEN = [
        'INSERT', 'UPDATE', 'DELETE', 'DROP', 'ALTER', 'CREATE', 'TRUNCATE',
        'RENAME', 'GRANT', 'REVOKE', 'REPLACE', 'MERGE', 'CALL', 'EXEC',
        'EXECUTE', 'PREPARE', 'DEALLOCATE', 'HANDLER', 'LOCK', 'UNLOCK',
        'LOAD', 'OUTFILE', 'DUMPFILE', r'INTO\s+OUTFILE', r'INTO\s+DUMPFILE',
        r'SET\s+GLOBAL', r'SET\s+SESSION', 'SHOW', 'DESCRIBE', r'DESC\s',
        'EXPLAIN', r'USE\s+', 'BENCHMARK', 'SLEEP',
    ]

    @classmethod
    def validate_and_prepare(cls, sql: str) -> str:
        sql = sql.strip().rstrip('; \t\n\r\x00')

        if not sql:
            raise ValueError('Empty SQL')

        if cls._contains_unquoted_semicolon(sql):
            raise ValueError('Multiple statements not allowed')

        stripped = cls._strip_comments(sql)

        if not re.match(r'^\s*(SELECT|WITH)\b', stripped, re.IGNORECASE):
            raise ValueError('Only SELECT / WITH queries are allowed')

        for kw in cls.FORBIDDEN:
            if re.search(r'\b' + kw + r'\b', stripped, re.IGNORECASE):
                raise ValueError(f'Forbidden keyword detected: {kw}')

        tables = cls._extract_tables(stripped)
        for t in tables:
            if t not in cls.ALLOWED_TABLES:
                raise ValueError(f'Disallowed table: {t}')

        return cls._enforce_limit(sql, cls.MAX_ROWS)

    @staticmethod
    def _contains_unquoted_semicolon(sql: str) -> bool:
        in_single = in_double = in_backtick = False
        prev = ''
        for ch in sql:
            if ch == "'" and prev != '\\' and not in_double and not in_backtick:
                in_single = not in_single
            elif ch == '"' and prev != '\\' and not in_single and not in_backtick:
                in_double = not in_double
            elif ch == '`' and not in_single and not in_double:
                in_backtick = not in_backtick
            elif ch == ';' and not in_single and not in_double and not in_backtick:
                return True
            prev = ch
        return False

    @staticmethod
    def _strip_comments(sql: str) -> str:
        sql = re.sub(r'/\*.*?\*/', ' ', sql, flags=re.DOTALL)
        sql = re.sub(r'--[^\n]*', ' ', sql)
        sql = re.sub(r'#[^\n]*', ' ', sql)
        return sql

    @classmethod
    def _extract_tables(cls, sql: str) -> list:
        # Collect CTE alias names — they are not real tables
        cte_names = set()
        for with_body in re.findall(r'\bWITH\s+(.+?)\s+SELECT\b', sql, re.IGNORECASE | re.DOTALL):
            for name in re.findall(r'\b([a-zA-Z_][a-zA-Z0-9_]*)\s+AS\s*\(', with_body, re.IGNORECASE):
                cte_names.add(name.lower())

        tables = []
        for t in re.findall(r'\b(?:FROM|JOIN)\s+`?([a-zA-Z_][a-zA-Z0-9_]*)`?', sql, re.IGNORECASE):
            lower = t.lower()
            if lower not in cte_names:
                tables.append(lower)
        return list(set(tables))

    @staticmethod
    def _enforce_limit(sql: str, max_rows: int) -> str:
        m = re.search(
            r'\bLIMIT\s+(\d+)(?:\s*(?:,|OFFSET)\s*(\d+))?\s*$',
            sql, re.IGNORECASE
        )
        if m:
            full_match = m.group(0)
            is_offset = 'OFFSET' in full_match.upper()
            has_comma = ',' in full_match

            count = int(m.group(1)) if not has_comma else int(m.group(2) or m.group(1))
            if is_offset:
                count = int(m.group(1))

            if count > max_rows:
                if is_offset:
                    sql = re.sub(
                        r'\bLIMIT\s+(\d+)\s+OFFSET\s+(\d+)\s*$',
                        f'LIMIT {max_rows} OFFSET \\2',
                        sql, flags=re.IGNORECASE
                    )
                elif has_comma:
                    sql = re.sub(
                        r'\bLIMIT\s+(\d+)\s*,\s*(\d+)\s*$',
                        f'LIMIT \\1, {max_rows}',
                        sql, flags=re.IGNORECASE
                    )
                else:
                    sql = re.sub(
                        r'\bLIMIT\s+(\d+)\s*$',
                        f'LIMIT {max_rows}',
                        sql, flags=re.IGNORECASE
                    )
            return sql

        return sql + f' LIMIT {max_rows}'
